src/backend/perf-events.c

```diff
@@ -322,22 +329,41 @@ int perf_events_init(int buf_size)
 		size += page_size;
 
 		perf_event_table[i].event->buf_size = size;
-		aligned_buf_size += size;
 
-		log_debug("%s using %u (%u pages) of buffer per CPU.\n",
+		if (perf_event_table[i].share_cpu) {
+			aligned_buf_size_cpu_share += size;
+		} else {
+			aligned_buf_size_cpu_exclusive += size;
+		}
+
+		log_debug("%s using %u (%u pages) of buffer%s.\n",
 				perf_event_table[i].event->name,
 				perf_event_table[i].event->buf_size,
-				perf_event_table[i].event->buf_size / page_size);
+				perf_event_table[i].event->buf_size / page_size,
+				perf_event_table[i].share_cpu ? "" : " per CPU");
 	}
 
-	log_debug("Perf buffer size aligned to %ld (%lxMB) per CPU.\n", aligned_buf_size, aligned_buf_size >> 20);
+	log_debug("Perf buffer size aligned to %ld (%lxMB) per CPU,"
+		" and %ld (%lxMB) of shared CPU\n",
+		aligned_buf_size_cpu_exclusive, aligned_buf_size_cpu_exclusive >> 20,
+		aligned_buf_size_cpu_share, aligned_buf_size_cpu_share >> 20);
 
 	trampo_page = malloc(page_size);
 
-	return perf_event_enabled_num;
+	return 0;
+}
+
+int perf_ring_setup_cpu_share(struct PerfEventRing *ring)
+{
+	return perf_ring_setup(ring, 0, -1);
```

Comment:
According to `man perf_event_open`

>       pid == 0 and cpu == -1
>              This measures the calling process/thread on any CPU.

it seems this will make memstrack measure itself? I think we should almost always be using -1, util we add the feature to measure a specified pid.

Reply:
> 



> According to `man perf_event_open`
> 
> > ```
> >   pid == 0 and cpu == -1
> >          This measures the calling process/thread on any CPU.
> > ```
The word 'calling' is misleading, I thought the calling means the process which actually triggered the perf event. However it can also mean the process which setup the perf event, so memstrack in this case...

> 
> it seems this will make memstrack measure itself? I think we should almost always be using -1, until we add the feature to measure a specified pid.

Reply:
Right... That's a problem indeed. Maybe we have to stick with the previous approach then, my concern was about the performance impact but with perf this seems no better way to me.

Reply:
OK, thanks for your comments, kasong, I will close this MR since the previous approach is merged. Thanks!